### DISCUSSION TITLE
Stop supporting Integrations that use GPIO in the core

### DISCUSSION THREAD

**Discussion by frenck** [home-assistant/architecture#596] [General] [answered]
## Summary

I want to make a bold suggestion, which boils down to a simple and single statement:

> No longer support integrations that rely on GPIO in the Home Assistant core.

## Proposal

We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).
Most of these integrations are low in use, often lack maintenance and above all, are generally better supported by projects like ESPHome.

Just some grabs from the stats:

![image](https://user-images.githubusercontent.com/195327/127837039-6295f869-1c68-4cd8-91a7-4ce2bfe1d896.png)

![image](https://user-images.githubusercontent.com/195327/127837182-44c8f8f7-75b8-4d7e-8c70-5ab9b11c6f3a.png)

![image](https://user-images.githubusercontent.com/195327/127837197-36cc8caf-30ec-4665-877b-4bbf8322c44d.png)

![image](https://user-images.githubusercontent.com/195327/127837238-1b60d3af-e681-4260-9372-c081a5d7fb9b.png)

![image](https://user-images.githubusercontent.com/195327/127837312-b4565819-2ad5-4732-b945-79325c72e7bc.png)
![image](https://user-images.githubusercontent.com/195327/127837341-7b0a207d-0753-4ffd-80e6-a8144010fe65.png)

![image](https://user-images.githubusercontent.com/195327/127837359-3ef24c0a-72ad-4097-88a7-724133bb6a3d.png)
![image](https://user-images.githubusercontent.com/195327/127837401-89f6e8fb-6e4a-430b-9158-9418da610f6d.png)

And we are getting new ones too, like here: <https://github.com/home-assistant/core/pull/53821>

My suggestion is that we no longer accept integrations communicating with GPIO for the following reasoning:

- The usage is low, but it adds the same amount of maintenance and review time to the core.
- Most of these integrations are unmaintained.
- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.
- The use of GPIO via containers (including our Home Assistant OS, Container & Supervised installation methods) is often unstable or complicated.
- Support for GPIO on devices, milage varies (looking at you Raspberry Pi on 64 bits platforms)

## Consequences

No longer accepting integration that integrates with these devices over GPIO that could be integrated (or are already integrated) by an ESPHome component.

**Comment by Misiu** [comment]
I disagree.
There are devices that are tight coupled with Raspberry Pi, like Argon ONE case (https://www.argon40.com/argon-one-m-2-case-for-raspberry-pi-4.html) that have active fan control via I2C. I've built a custom component to control it (https://github.com/Misiu/argon40) and I'm in process of rewriting it to support fan entity and config flow. When I finish I want to add it to the core.
The same applies to PI HATs, like for example https://wiki.52pi.com/index.php/DockerPi_4_Channel_Relay_SKU:_EP-0099 or https://sequentmicrosystems.com/collections/all-io-cards/products/raspberry-pi-relays-heavy-duty-hat.
You can't use ESPHome to control them, because they need to be attached to the PI.

**Reply by balloob** [reply]
These are rare occurrences and it's still possible as a custom component.

**Comment by Jc2k** [comment]
In favour of removing

* anything thats unmaintained or just plain broken
* eventually removing things with no users.
* has a code owner but the integration is holding up removing a deprecated API, code owner has no plans to resolve.

And that applies generally to all integrations. E.g. AIUI the dyson fan integration which i believe is completely broken for everyone at this point.

I do wonder if some of those integrations could merge to reduce maintenance burden. I imagine the 3 bosch sensors are very similar?

I don't support removing things that have active code owners or blocking new integrations (especially but not limited to those from known contributors). These might be the contributors that with some mentoring could help with the code review backlog.

Maybe there has to be a popular request thread for it on the forums as a threshold to be added? Or an existing custom component with lots of users. Both of these tests would have prohibited one of my own integrations from getting added, even though it is an incredibly simple project, has since found users, and was submitted by a contributor with a track record for fixing his bugs.

**Reply by thecode** [reply]
In general I am in favor of not supporting things which rely on dedicated HW which runs the core, such as GPIOs/I2C/SPI since this are platform depended and we may end up creating a broken integration.
However after reading @Jc2k comment I have doubts if we should remove active & maintained integrations, If an integration has an owner and he is maintaining it and adapting it to different platforms it might not be a good decision to drop them.
I think we should start by notifying owners for unmaintained integrations about the removal and by this decide what to remove or not. This applies also to other broken integrations. In theory even if integration is removed it can be re-added in the future if someone wants to take ownership and maintain it.
I suggest to start with this, we may anyhow end up in removing all GPIO related integrations just in a slower process :-)

**Comment by balloob** [comment]
I've personally never been a fan of the GPIO ones. It's an edge use case to have sensors attached to your Home Assistant instance as people tend to put their HA devices in the closet. Attaching them to a $2 ESP device with ESPHome is the way to go. And if there are users that want to use them, they can always rely on custom components for this.

We are a home automation and energy management platform.

**Comment by aSauerwein** [comment]
Even though #53821 is my PR I understand your point that you no longer want to accept  code for GPIOs to Core. 

Nevertheless I thought that is not a rare use-case. For myself I install RPIs with Homeassistant in remote cottages to turn on heating and measure temperature (onewire).  And for this cases I like to use the GPIOs that are already attached instead of ESPHome. 

For Integrations like my PCF8574 it would be great if this are searchable on some Home Assistant Marketplace and install them 1-click to custom_components. 
Like Grafana Dashboards, Ansible Galaxy etc. 
This way every User can install components (easily) at their own risk and you do not have to maintain them. 

Sorry if you already have that, and I just did not find it.

**Reply by balloob** [reply]
There is https://hacs.xyz/

**Reply by joakimeriksson** [reply]
I agree with this - from my perspective home assistant is the home automation platform and I guess the core repository should focus on these parts - rather than low-level device drivers (such as GPIO). Is there a way to move these integrations into other repositories - for HA low-level "drivers"?

**Comment by frenck** [comment, marked as answer]
Opened up PR to enforce it: <https://github.com/home-assistant/architecture/pull/686>

**Comment by Torok42** [comment]
Very bad idea based on incomplete data

For exemple : 
1-wire (231 installations)
![image](https://user-images.githubusercontent.com/58152695/147789760-e85d0b7a-c66b-44ec-b68a-c6bd53258f9d.png)

DHT (180 installations)
![image](https://user-images.githubusercontent.com/58152695/147789801-24994a33-46c0-4343-b91c-bb5189e60ef6.png)

There are still several people who use them.....

**Reply by frenck** [reply]
Those are not high numbers either, sure people use them, the point is not just usage.

The generally really low(er) usage is just one of the points made, but not the sole factor.

Anyways, we won't be making efforts in blocking these type of integrations and they still can be offered and maintained as custom integrations as well.

**Comment by Misiu** [comment]
Does that mean that all integrations relying on GPIO will be deprecated? For example `RPI_GPIO`, `ORANGEPI_GPIO`, `RPI_RF`, and others?
There are plenty of integrations that rely on packages that communicate directly via GPIO (https://github.com/home-assistant/core/blob/dev/requirements_all.txt, search for `GPIO`).

The decision has been made, I don't argue with it, but maybe give the users an easier way to add them. I know that all of those integrations can be migrated to HACS, but I needed so much time when I installed HACS for the first time and when I explained that to my friends. It is easier these days, but still not that easy. If we could set up HACS directly from HA (like we install addons or enable integrations) then we all would be happy. Maybe HACS could be a part of Home Assistant 🤩?

**Reply by frenck** [reply]
> Does that mean that all integrations relying on GPIO will be deprecated? 

Yes

> There are plenty of integrations that rely on packages that communicate directly via GPIO

Yes, I'm in the process of marking them deprecated as we speak.

The rest of your message is not really related to this ADR and more a feature request for the community forum.

**Reply by Misiu** [reply]
> The rest of your message is not really related to this ADR and more a feature request for the community forum.

Will do that in a second.

**Comment by pascallj** [comment]
Although I agree that the GPIO integrations should be more uniform and one standard should have existed (I was working on a universal gpiochip based solution, but I guess I'll scratch that then), I think this is rather harsh. 

* A lot of other integrations are unmaintained as well, which is often the case in open source projects. If it ain't broke, there is no motivation to fix or enhance it.
* ESPHome is not an alternative at all in this case. People who are using GPIO on their Home Assistant devices, don't seem to have a reason for extra devices. Apparently what they want to control is in the vicinity of their server. For example my home alarm and door sensor are right next to my modem and therefore my HASS server. Why should I need to put an extra device in between those?
* People who want to tinker with the direct GPIO, are often skilled enough to make it work in container environments if they so desire. But I think that most of the people who use these integrations are running Core anyway. 
* Different support on different devices would have been solved with the integration I have been working on.

However, I can see the idea of extracting the 'backend' out of HASS. In this case a more promising alternative could be [MQTT IO](https://github.com/flyte/mqtt-io).

**Reply by balloob** [reply]
The integrations are removed from core, not banned. So if someone puts them on HACS a user can install them again.

Something like MQTT-IO could also be made into an add-on.

**Comment by Yevgenium** [comment]
The Raspberry Pi platform is officially supported by Home Assistant and is recommended for use. GPIO is one of the main hardware features of the Raspberry Pi.
According to analytics, the Raspberry Pi GPIO integration is currently used by more than 1,100 people, which indicates that it is quite in demand.
I hope we could keep at least the basic GPIO capabilities that this integration currently provides.

**Reply by frenck** [reply]
> I hope we could keep at least the basic GPIO capabilities that this integration currently provides.

It won't. The decision has been made and formalized in #686.
The integration has been deprecated, as of Home Assistant Core 2022.2, and pending removal for Home Assistant Core 2022.6.

For more information and the preliminary announcement, you can already read the breaking changes section in the Home Assistant Core beta release notes on https://rc.home-assistant.io

**Reply by shbatm** [reply]
> I hope we could keep at least the basic GPIO capabilities that this integration currently provides.

@Yevgenium if you want a suggestion on a replacement method, [MQTT-IO](https://github.com/flyte/mqtt-io) has been working well for me, may be a good candidate for an Add-on.

**Comment by Hedda** [comment]
> I want to make a bold suggestion, which boils down to a simple and single statement:
> 
> > No longer support integrations that rely on GPIO in the Home Assistant core.

Just to clarify, this does not include UART/serial via GPIO for integrations with radio modules using the UART serial port via GPIO header, or?

That is for example used for Zigbee Coordinator radio shields/hats (for the Raspberry Pi GPIO header) compatible with the [ZHA integration:

https://www.home-assistant.io/integrations/zha#compatible-hardware

Examples of specific Zigbee radio serial / UART modules connected via Raspberry Pi GPIO header:

* Home Assistant Yellow onboard MGM210P module https://www.crowdsupply.com/nabu-casa/home-assistant-yellow
* Electrolama zoe2 HAT https://electrolama.com/projects/zoe2/
* Electrolama zoe HAT https://electrolama.com/projects/zoe/
* ZigStar Shield https://zig-star.com/projects/zigbee-shield/
* cod.m Zigbee CC2652P RPi Module https://shop.codm.de/automation/zigbee/33/zigbee-cc2652p2-raspberry-pi-module
* GiovanniCas Zigbee PiHat https://www.tindie.com/products/GiovanniCas/zigbee-hat-with-cc2538-for-raspberry/
* RaspBee II (a.k.a. RaspBee 2) Raspberry Pi Shield from dresden elektronik https://www.dresden-elektronik.com/product/raspbee-II.html
* ZiGate PiZiGate+ https://zigate.fr/produit/pizigatev2/
* Elelabs Zigbee Raspberry Pi Shields https://elelabs.com/products/elelabs_zigbee_shield.html
  * Elelabs Zigbee Raspberry Pi Shield ELR023 (EFR32MG13P) 
  * Elelabs Zigbee Raspberry Pi Shield ELR022 (EFR32MG1B based)
  * Elelabs Zigbee Raspberry Pi Shield EZBPIS (EM357 based)
* POPP ZB-Shield (701561) https://www.popp.eu/zb-shield/

Many other integrations also support radio shields/HATs using UART RX TX GPIO pins, like; EnOcean, Z-Wave JS, RFLink, XBee(?). Examples:

* Aeotec Z-Pi 7 https://aeotec.com/z-wave-home-automation/development-kit-pcb.html
* z-wave.me RaZberry 7 sheild https://z-wave.me/products/razberry/
* z-wave.me RaZberry 2 / RaZberry 5

and

* Joy-it EnOcean Pi https://www.conrad.com/p/joy-it-enocean-pi-breadboard-shield-1000221

I believe future integrations using radio shields/HATs using UART via GPIO may include radios for Matter/Thread (OpenThread) and LoRa(?).

**Reply by frenck** [reply]
Please read the ADR and your question will be answered

**Reply by Hedda** [reply]
Sorry but ADR specifically only mention SPI and i2c busses and not specifically serial using UART via GPIO which is why I am asking.

Will existing/ future integrations be able to use radio serial shields/HATs using UART RX TX GPIO pins like ZHA integration does today?

**Reply by frenck** [reply]
![image](https://user-images.githubusercontent.com/195327/152300860-afeee1c8-2576-4694-a623-789a947c7821.png)

**Comment by RubenKelevra** [comment]
This decision should be reverted.

I don't think that anybody was part of this discussion who is actively using it - as there was no hint in any blog entry etc. about this major feature cut.


Not only are there many cooling solutions using this, but there are UPS additions etc.

If you proceed with this feature cut I wouldn't consider the Raspberry PI an officially supported platform.

So please go ahead and announce this in the blog and have a look at the response. :)

**Reply by Hedda** [reply]
> So please go ahead and announce this in the blog and have a look at the response. :)

FYI, an announcement is listed as a part among "Breaking Changes" in Home Assistant 2022.2 release notes on the blog:

https://www.home-assistant.io/blog/2022/02/02/release-20222/

https://www.home-assistant.io/blog/2022/02/02/release-20222/#breaking-changes

https://community.home-assistant.io/t/2022-2-lets-start-streamlining/387648/2

**Reply by RubenKelevra** [reply]
@Hedda right, but this was published around 2 month AFTER the decision was made. 

So how many people from the community were notified BEFORE the decision was made at 20th of December?

And since the only explanation is, that there's too little use. How many of the other 764 lower used integration will be removed in the future?

([Analytics](https://analytics.home-assistant.io/#integrations) show Rasberry Pi's GPIO integration at place 122 of 886 integrations).

This decision is just wrong and should be reverted.

**Comment by thecode** [comment]
I believe that integrations that interact directly with low level hardware should not be a part of Home Assistant Core, it makes it impossible to maintain and test them. Maintaining those as external integrations removes the coupling between core releases and integration releases and make it easier to be flexible to different hardware types.

Since there are users for the `rpi_gpio` integration I will maintain it as custom integration and will update changes to allow it to run on future core releases. You can find the custom `rpi_gpio` integration here: [thecode/ha-rpi_gpio](https://github.com/thecode/ha-rpi_gpio) or add it through [HACS](https://hacs.xyz/).

**Reply by Vrobenmat** [reply]
Hey! this is great! will this also support remote_rpi_gpio or is that integration remaining in core? e.g. could I run pigpiod on localhost and just use remote_rpi_gpio?

**Reply by thecode** [reply]
> Hey! this is great! will this also support remote_rpi_gpio or is that integration remaining in core?

The `remote_rpi_gpio` does not interact directly with GPIOs and is not deprecated.

**Reply by Vrobenmat** [reply]
Now is gets interesting. I have one these massive [relay boards](https://www.waveshare.com/product/raspberry-pi/hats/motors-relays/rpi-relay-board-b.htm) in my fusebox running HA and, amongst other things, controlling my heating system so local control is a must - will play around to see if I prefer pigpiod or HACS as a stable alternative. Definitely don't want to go the MQTT route just for this.

**Reply by Siggin0er** [reply]
Aloah!
Great news that you maintain it as a custom_component!
I also use / need the mcp23017 integration as well, could you implement that, too? Unfortunately I have no idea of maintaining such things on my own, even if I wish i could. Perhaps it's not that difficult, but I don't have time to dig into it on my own. Programming should not be a problem...

**Reply by frenck** [reply]
> I also use / need the mcp23017 integration as well, could you implement that, too? 

Let move such requests to the custom integration repository instead 😉 

Thanks!

**Comment by BradleyFord** [comment]
I noticed the change, and I don't even submit any code. But the reason is
quite ok for me.
If it is really that big a deal I guess you could fork the project?

**Comment by RubenKelevra** [comment]
Interesting, how did you get notified of this discussion prior to the 20th of December?

It's not a big of a deal on its own - sure I and probably thousands of other people have to touch their setup or throw away some hardware and replace it with something new (or use the new custom component).

The question for me is: What's next? There are hundreds off less used components which could be removed in the same fashion with the same rationale behind it as 1% of the userbase which is supplying statistical data doesn't seem to be enough to care about.

Additionally the Raspberry PI is officially supported hardware, so I'm wondering if this is dropped is this just the start? Will other hardware specific features be dropped as well?

And I don't think it's fair on a community driven project to not have at least a discussion or a open vote for such major decisions. 

Instead of publishing two month later a changelog entry and if there are concerns raised to reply with "the decision was formalised" and of discussion, basically.

No, this isn't a personal thing. Please don't understand it this way. I just wonder why dropping support isn't handled better in the Home Assistant project. I would expect such decision making from a company like Apple or Samsung which just drop support for one of their products and the users have to buy something new - not from a community driven project like HA.

Why doesn't HA put a section in the Changelog where people can vote on such decisions?

"We noticed that we use a lot of development time on X, and we like to drop the support on X in N versions to streamline the development process.

Do you think this is the right decision?

- Yes
- No
- Unsure

How will this change affect you?

- I don't use it / don't plan to use it
- I plan to use it
- I have it configured but don't actively use it
- It would make some of my hardware obsolete / need to replace it
- I could change my setup easily to avoid using it

If you want to discuss this further, join our discussion here"

Followed by a link to this thread.

**Reply by aidbish** [reply]
> Why doesn't HA put a section in the Changelog where people can vote on such decisions?

Because only those that care for that option would vote, the rest wouldn't bother so the result would always go in favour to the people who voted whether its the correct decision or not.

**Comment by norbertohenriques** [comment]
I do use the rpi_gpio platform to integrate my Jablotron alarm system with HASSIO, as I have a Jablotron binary output module that has several outputs linked to the alarm sensors around the house.
This enables me to use the alarm sensors as movement sensors to trigger light switches, for example.
I'm very sorry to know that there will be no official support for this platform and that we'll have to rely on 3rd party integrations via HACS in order to keep it working.

**Comment by dtext** [comment]
I run homeassistant in docker on a raspberry pi 4. There is a Bosch BME280 sensor connected to it via i2c. Using the now deprecated `bme280` integration to read values from the `/dev/i2c-1` device which i mounted into the homeassistant container. I don't see any issues here.

Please explain how this is "harder" or in some way "worse" than buying an ESP32, power supply, setting everything up with ESPHome and connecting yet another device to my network, adding yet another moving part to my home automation setup. I'm genuinely interested in the reasoning behind this.

I realise that is not your only argument for dropping support, and ultimately it's your decision. However, I have another sidenote on the usage numbers: I have all analytics turned off. Having them on kind of defeats the purpose of self hosting to me. I would imagine that many home assistant users actually have them turned off, which _could_ mean that your usage statistics are way off (if they acually come from analytics data).

Please consider offering a clear migration path for people who do happen to be using gpio (which is one of the pi's central features if you ask me). For example, instead of just removing them, you could move them to new repositories, maybe even search for maintainers, etc.. I am quite disappointed with the decision as it is, because it will completely break my setup in 2022.6 without a clear path to move on for me.

**Reply by thecode** [reply]
Just one note about analytics, when looking at numbers the logic is to look at the percentage, assuming that a statistics of 100K users is large enough to be accurate.  If it is not the case and people who use specific integration are all belong to a group who decide not to share analytics than something is wrong here. I recommend enabling analytics and I don't think it defeats the purpose of self hosting. My setup is self hosted, not only Home Assistant and I still share analytics. If you also think the analytics is not representing the real state you have an option to change it by sharing analytics.

**Reply by dtext** [reply]
Thanks for your reply!

>If it is not the case and people who use specific integration are all belong to a group who decide not to share analytics than something is wrong here.

My thoughts on this, more or less. Let's assume that advanced people turn off analytics, because they actually bothered to tinker with their installation. (Not saying that **all** advanced people do this.) Wouldn't they also be the ones who connect sensors to their device via i2c and who are knowledgeable enough to get it to work in docker? Just a theory, of course.
There are other (to me: preferable) ways to get community feedback on big steps before actually taking them. This feels like quite a big step to me. One of them is to have a discussion like this, which is why I decided to share my views here :)

However, I don't want to enable analytics, and it wouldn't really change anything about this decision, would it?

**Comment by TheFax** [comment]
Removing this integration will give me problems.
I am really disappointed. All right. I am aware that free software cannot be perfect.

### LEARN
**But it is important that this event teaches us one thing**: integrations must born with the idea of ​​being almost eternal (created - tested - maintained) and not simply created for the need of a good programmer who had the skills and a single device to be made compatible in its home.
So I ask myself: why are there 454 integrations with less than 100 users, 155 integrations with less than 10 users and 56 integrations with 1 or 0 users? Do we consider these integrations useful and stable over time? **But above all**, are we informing the new possible future Home Assistant users of this? Are we encouraging or discouraging to use them? 

The same goes for the other components / parts of HA.

With all due respect, but before I buy a card or device with my money, I make sure that it fits my needs and that it can last as long as possible.
I did this when I bought my router and I installed OpenWRT.
I did this when I bought Raspberry to install Home Assistant on it. And to do this I read what was wrote on your website at the time.

### HACS
I have read that you can use HACS to solve. Ok. But as soon as I hit the "Download" button on the website, I am immediately slapped by a message that says:

> "_**CAUTION**! If all of the points in the list below does not match you, **do not try to setup** / use HACS._".

WHY? Is it dangerous? Can it break other things in my Home Assistant installation? What is a "device OAuth flow for authentication against the GitHub API"? Will I am able to configure it? Is HACS an official component for Home Assistant or is it manatained only by a single programmer that will abandon it / make it NOT FREE in one or two years?  The feeling I get is: fear.

### CONCLUSION
The keyword of my intervention here is: **stability over time**. This is the only thing I ask.

![immagine](https://user-images.githubusercontent.com/19174767/175805223-470df514-5bdf-4728-abab-c1b6c38780e0.png)

**Reply by electromost** [reply]
I use all interfaces - UART (RS485), 1-wire, I2C and GPIO inputs / outputs. Now it has all stopped working. It was possible to restore only the inputs / outputs and 1-wire.
I'm disappointed in Home Assistant. It is very easy to break something and give nothing in return.

**Reply by frenck** [reply]
We removed them from the core, but do not block them as custom integrations. Some of them are now available as a custom integration (e.g., GPIO is available via HACS).

This decision has been made and is very unlikely to be reverted.

Again, all that changed, is that these are no longer maintained in core. But can still be created and maintained as a custom integration.

Locking this discussion, as the architectural decision has been resolved.
